File: render/figures/walkthrough.py
"""The month on one axis, with the event rail under it."""

from __future__ import annotations

import pandas as pd
import plotly.graph_objects as go

from copytext import t, tpl

from .. import theme
from ..theme import MONO

# ---------------------------------------------------------------------------

#: (column, unit, colour slot, dash pattern, rules that read it)
#:
#: Divided by the maximum rather than rescaled min-max because zero has to stay
#: zero: for user A, "no late-night minutes" is the finding, and min-max would
#: paint it halfway up the axis.
#:
#: The dash pattern is a second encoding, not decoration. The worst adjacent
#: pair in this palette sits at ΔE 10.3 on this surface, which is inside the
#: floor band for protan colour blindness.
TRACKED = [
    ("night_min", "min", 0, "solid", "night_drift · night_streak"),
    ("night_end_min", "min", 1, "dash", "night_drift"),
    ("screen_min", "min", 2, "solid", "screen_jump"),
    ("longest_offline_h", "h", 3, "dot", "offline_record"),
    ("blocks", "", 4, "dashdot", "calm_week"),
    ("blocks_sensitive", "", 5, "dash", "sensitive_spike · filter_calm"),
    ("distract_pct", "%", 6, "dot", "focus_week"),
]

#: Series visible on open. The rest come in with a legend click: seven lines at
#: once do not read, and starting with all of them on forces the reader to
#: switch things off rather than on.
TRACKED_DEFAULT = {"night_min", "night_end_min", "screen_min"}

#: Height of the event rail, below the data zero.
_RAIL = -9


def _derive_tracked(df: pd.DataFrame) -> pd.DataFrame:
    """Derived columns that exist only for this chart."""
    df = df.copy()
    # The time of the last screen is expressed as minutes past 23:00, so zero
    # means "went dark on time" rather than "midnight".
    df["night_end_min"] = (df["night_end_h"] - 23) * 60
    df["longest_offline_h"] = df["longest_offline_s"] / 3600
    df["distract_pct"] = df["distract_share"] * 100
    return df


def tracked_series(df: pd.DataFrame, user: str, cursor,
                   nudge_days: set, alert_days: dict, positive_days: dict,
                   h: int = 560):
    """Every watched variable on one axis, each as a share of its own maximum.

    Below zero, an event rail sharing the time axis with the data that
    explains it.
    """
    d = _derive_tracked(df)
    fig = go.Figure()

    for i, (col, unit, slot, trazo, reglas) in enumerate(TRACKED):
        label = t(f"tracked.{col}")
        serie = d[col]
        tope = serie.max()
        if pd.isna(tope) or tope <= 0:
            # Series flat at zero: drawn anyway, hugging the axis, so it is
            # visible that the data exists and is zero.
            norm = serie.fillna(0) * 0
            note_ = t("series.no_activity")
        else:
            norm = serie / tope * 100
            note_ = ""
        fig.add_trace(go.Scatter(
            x=d["day"], y=norm, mode="lines+markers",
            name=f"{label}{note_}",
            # The group only titles the legend. With
            # `groupclick="toggleitem"` each entry toggles on its own; without
            # that option, Plotly switches the whole group off in one click.
            legendgroup="datos",
            legendgrouptitle=(dict(text=t("legend.tracked"))
                              if i == 0 else None),
            visible=True if col in TRACKED_DEFAULT else "legendonly",
            line=dict(color=theme.CATEGORICAL[slot], width=2, dash=trazo),
            marker=dict(size=5, color=theme.CATEGORICAL[slot],
                        symbol=["circle", "square", "diamond", "cross",
                                "x", "triangle-up", "pentagon"][slot]),
            customdata=serie,
            hovertemplate=tpl("hover.tracked", label=label, unit=unit,
                              rules=reglas),
        ))

    # --- event rail ---------------------------------------------------------
    events = [
        (t("event.nudge"), sorted(nudge_days), "circle", theme.WARN,
         t("event.nudge.detail")),
        (t("event.alert"),
         sorted(k for k, v in alert_days.items() if v == "sent"),
         "triangle-up", theme.SERIOUS, t("event.alert.detail")),
        (t("event.digest"),
         sorted(k for k, v in alert_days.items() if v == "summary"),
         "diamond", theme.INK_2, t("event.digest.detail")),
        (t("event.positive"), sorted(positive_days), "star", theme.GOOD,
         t("event.positive.detail")),
    ]
    for i, (name, days, symbol, color, detail) in enumerate(events):
        fig.add_trace(go.Scatter(
            x=days, y=[_RAIL] * len(days), mode="markers", name=name,
            legendgroup="events",
            legendgrouptitle=(dict(text=t("legend.emissions"))
                              if i == 0 else None),
            marker=dict(symbol=symbol, size=11, color=color,
                        line=dict(color=theme.CARD, width=1)),
            hovertemplate=tpl("hover.event", detail=detail),
            showlegend=True,
        ))

    fig.add_hline(y=0, line=dict(color=theme.RULE, width=1))
    fig.add_vline(x=cursor, line=dict(color=theme.INK, width=2))

    fig.update_layout(
        height=h, hovermode="x unified",
        yaxis_title=t("axis.pct_of_max"),
        margin=dict(t=44, r=24, b=120, l=64),
        legend=dict(orientation="h", y=-0.24, x=0, xanchor="left",
                    yanchor="top", groupclick="toggleitem",
                    font=dict(family=MONO, size=11, color=theme.INK_2),
                    grouptitlefont=dict(family=MONO, size=11, color=theme.MUTED)),
    )
    fig.update_yaxes(range=[_RAIL - 5, 108], dtick=25,
                     tickvals=[0, 25, 50, 75, 100],
                     ticktext=["0", "25", "50", "75", t("axis.pct_max_tick")])
    fig.update_xaxes(tickformat="%d %b")
    fig.add_annotation(xref="paper", x=0, y=_RAIL, yanchor="middle",
                       xanchor="right", xshift=-8, text=t("annotation.events"),
                       showarrow=False,
                       font=dict(family=MONO, size=10, color=theme.MUTED))
    return fig

File: web/charts/walkthrough.ts
/**
 * The month on one axis, with the event rail under it.
 */

import { annotation, axis as unit, hover } from "../copy/figures";
import { MONO, type Surface } from "../theme";
import type { DailyRow, Day, ReplayDay } from "../types/index";
import { frame } from "./frame";
import type { Figure, Trace } from "./plotly";

/**
 * Every variable the alert rules read.
 *
 * Divided by its own maximum rather than rescaled min-max, because zero has
 * to stay zero: for user A, "no late-night minutes" is the finding, and
 * min-max would paint it halfway up the axis.
 *
 * The dash pattern is a second encoding, not decoration. The worst adjacent
 * pair in this palette sits at ΔE 10.3 on this surface, inside the floor band
 * for protan colour blindness.
 */
const TRACKED = [
  { key: "night_min", label: "Late-night screen", unit: "min", slot: 0,
    dash: "solid", symbol: "circle", rules: "night_drift · night_streak" },
  { key: "night_end_min", label: "Last screen (from 23:00)", unit: "min",
    slot: 1, dash: "dash", symbol: "square", rules: "night_drift" },
  { key: "screen_min", label: "Screen per day", unit: "min", slot: 2,
    dash: "solid", symbol: "diamond", rules: "screen_jump" },
  { key: "longest_offline_h", label: "Longest break", unit: "h", slot: 3,
    dash: "dot", symbol: "cross", rules: "offline_record" },
  { key: "blocks", label: "Blocks per day", unit: "", slot: 4,
    dash: "dashdot", symbol: "x", rules: "calm_week" },
  { key: "blocks_sensitive", label: "Sensitive attempts", unit: "", slot: 5,
    dash: "dash", symbol: "triangle-up", rules: "sensitive_spike · filter_calm" },
  { key: "distract_pct", label: "Distraction share", unit: "%", slot: 6,
    dash: "dot", symbol: "pentagon", rules: "focus_week" },
] as const;

/** Visible on open. Seven lines at once do not read, and starting with all of
 *  them on forces the reader to switch things off rather than on. */
const DEFAULT_ON = new Set(["night_min", "night_end_min", "screen_min"]);

/** Height of the event rail, below the data zero. */
const RAIL = -9;

/** The columns that exist only for this chart. */
function derived(row: DailyRow): Record<string, number | null> {
  return {
    // The last screen is minutes past 23:00, so zero means "went dark on
    // time" rather than "midnight".
    night_end_min: row.night_end_h === null ? null : (row.night_end_h - 23) * 60,
    longest_offline_h: row.longest_offline_h,
    distract_pct: row.distract_share * 100,
    night_min: row.night_min,
    screen_min: row.screen_min,
    blocks: row.blocks,
    blocks_sensitive: row.blocks_sensitive,
  };
}

/** Which days carried which emission. */
function rails(replay: ReplayDay[]) {
  const nudges: Day[] = [];
  const alerts: Day[] = [];
  const digests: Day[] = [];
  const positives: Day[] = [];
  for (const day of replay) {
    if (day.nudge?.fired) nudges.push(day.day);
    if (day.alert) alerts.push(day.day);
    else if (day.digest_entry) digests.push(day.day);
    if (day.positives.length) positives.push(day.day);
  }
  return { nudges, alerts, digests, positives };
}

export function trackedSeries(s: Surface, rows: DailyRow[],
                              replay: ReplayDay[], cursor: Day): Figure {
  const days = rows.map((r) => r.day);
  const values = rows.map(derived);
  const data: Trace[] = [];

  TRACKED.forEach((track, i) => {
    const raw = values.map((v) => v[track.key] ?? null);
    const top = Math.max(...raw.map((v) => v ?? 0));
    // A series flat at zero is drawn anyway, hugging the axis, so it is
    // visible that the data exists and is zero.
    const flat = !Number.isFinite(top) || top <= 0;
    const colour = s.categorical[track.slot] ?? s.ink;
    data.push({
      type: "scatter", mode: "lines+markers", x: days,
      y: flat ? raw.map(() => 0) : raw.map((v) => (v === null ? null : v / top * 100)),
      name: `${track.label}${flat ? " · no activity" : ""}`,
      // The group only titles the legend; `groupclick: "toggleitem"` keeps a
      // click on one entry from switching off the whole group.
      legendgroup: "tracked",
      ...(i === 0 ? { legendgrouptitle: { text: "Watched variables" } } : {}),
      visible: DEFAULT_ON.has(track.key) ? true : "legendonly",
      line: { color: colour, width: 2, dash: track.dash },
      marker: { size: 5, color: colour, symbol: track.symbol },
      customdata: raw,
      hovertemplate: hover.tracked(track.label, track.unit, track.rules),
    });
  });

  const { nudges, alerts, digests, positives } = rails(replay);
  const events = [
    { name: "Night with a nudge", days: nudges, symbol: "circle",
      color: s.warn, detail: "Night nudge on the device" },
    { name: "Alert", days: alerts, symbol: "triangle-up", color: s.serious,
      detail: "Notification shown on the phone" },
    { name: "Summary entry", days: digests, symbol: "diamond", color: s.ink2,
      detail: "Signal held for the weekly summary" },
    { name: "Reinforcement", days: positives, symbol: "star", color: s.good,
      detail: "Reinforcement sent" },
  ];
  events.forEach((event, i) => {
    data.push({
      type: "scatter", mode: "markers", x: event.days,
      y: event.days.map(() => RAIL), name: event.name,
      legendgroup: "events",
      ...(i === 0 ? { legendgrouptitle: { text: "Emissions" } } : {}),
      marker: { symbol: event.symbol, size: 11, color: event.color,
                line: { color: s.card, width: 1 } },
      hovertemplate: hover.event(event.detail),
      showlegend: true,
    });
  });

  return frame(s, data, {
    hovermode: "x unified",
    yaxis: { title: { text: unit.pctOfMax }, range: [RAIL - 5, 108],
             dtick: 25, tickvals: [0, 25, 50, 75, 100],
             ticktext: ["0", "25", "50", "75", "100 %"] },
    xaxis: { tickformat: "%d %b" },
    margin: { t: 44, r: 24, b: 120, l: 64 },
    legend: { orientation: "h", y: -0.24, x: 0, xanchor: "left",
              yanchor: "top", groupclick: "toggleitem",
              font: { family: MONO, size: 11, color: s.ink2 },
              grouptitlefont: { family: MONO, size: 11, color: s.muted } },
    shapes: [
      { type: "line", xref: "paper", x0: 0, x1: 1, y0: 0, y1: 0,
        line: { color: s.rule, width: 1 } },
      { type: "line", yref: "paper", y0: 0, y1: 1, x0: cursor, x1: cursor,
        line: { color: s.ink, width: 2 } },
    ],
    annotations: [{ xref: "paper", x: 0, y: RAIL, yanchor: "middle",
                    xanchor: "right", xshift: -8, text: annotation.events,
                    showarrow: false,
                    font: { family: MONO, size: 10, color: s.muted } }],
  }, 560);
}

/** The cursor is the only thing that moves when the day slider does. */
export function cursorUpdate(day: Day): Record<string, unknown> {
  return { "shapes[1].x0": day, "shapes[1].x1": day };
}
